client/src/controllers/SyncController.ts

```diff
@@ -0,0 +1,155 @@
+
+    this.getTargetElements().forEach((target) => {
+      const event = this.dispatch('start', {
+        target,
+        detail: { element: this.element, value: this.element.value },
+        cancelable: true,
+        bubbles: false,
+      });
+
+      if (event.defaultPrevented) {
+        allChecksPass = false;
+      }
+    });
+
+    this.disabledValue = allChecksPass;
+  }
+
+  apply() {
+    if (this.disabledValue) return;
+
+    const slugifiedTitle = cleanForSlug(this.element.value.trim(), true);
+
+    this.getTargetElements().forEach((target) => {
+      const event = this.dispatch('apply', {
+        target,
+        detail: { element: this.element, value: this.element.value },
+        bubbles: false,
+        cancelable: true,
+      });
+
+      if (event.defaultPrevented) return;
+      // eslint-disable-next-line no-param-reassign
+      target.value = slugifiedTitle;
+    });
+  }
+
+  check() {
+    let allChecksPass = true;
+
+    this.getTargetElements().forEach((target) => {
+      const event = this.dispatch('check', {
+        target,
+        detail: { element: this.element, value: this.element.value },
+        cancelable: true,
+        bubbles: false,
+      });
+
+      if (event.defaultPrevented) {
+        allChecksPass = false;
+      }
+    });
+
+    this.disabledValue = allChecksPass;
+  }
+
+  clear() {
+    if (this.disabledValue) return;
+
+    this.getTargetElements().forEach((target) => {
+      if (!target.value) return;
+      const event = this.dispatch('clear', {
+        target,
+        detail: { element: this.element, value: this.element.value },
+        cancelable: true,
+        bubbles: false,
+      });
+
+      if (event.defaultPrevented) return;
+      // eslint-disable-next-line no-param-reassign
+      target.value = '';
+    });
+  }
+
+  getTargetElements() {
+    return document.querySelectorAll<HTMLInputElement>(this.targetValue);
+  }
+
+  ping() {
+    if (this.disabledValue) return;
+
+    this.getTargetElements().forEach((target) => {
+      if (!target.value) return;
+
+      const event = this.dispatch('ping', {
+        target,
+        detail: { element: this.element, value: this.element.value },
+        cancelable: true,
+        bubbles: false,
+      });
+
+      if (event.defaultPrevented) return;
+
+      this.dispatch('change', {
+        target,
+        detail: { element: this.element, value: this.element.value },
+        cancelable: true,
+        bubbles: false,
+        prefix: '',
+      });
+    });
+  }
+
+  disconnect(): void {
+    let allChecksPass = true;
+
+    this.getTargetElements().forEach((target) => {
+      const event = this.dispatch('stop', {
+        target,
+        detail: { element: this.element, value: this.element.value },
+        cancelable: true,
+        bubbles: false,
+      });
+
+      if (event.defaultPrevented) {
+        allChecksPass = false;
+      }
+    });
+
+    this.disabledValue = allChecksPass;
+  }
+}
```

Comment:
ok, this is getting close but we now have a bit of a problem with lots of duplicated code, also I think the `'change'` event will actually be more useful as a default on any change (apply/clear).

Finally, we should not have to bring in the `cleanForSlug` util at all for this behaviour to work, that's the whole purpose of the `check`/`compare` `preventDefault` cycle.

Essentially, the sync field checks with it's target field if it's OK to be enabled, if that check fails (event.preventDefault() called), then the sync controller goes into disabled mode and does nothing else.

Here is a proposed approach.

* General handling of 'get me a bunch of elements that did not prevent default' is now in a renamed `getTargetElements`
* `ping` - this event bubbles, this is intentional.
* new value `quiet`, default false, if true, `change` events will NOT be dispatched when changes take place, note that this behaviour will not impact the slug field as it prevents default on apply so the value update / change dispatch will never happen.
* Using the `delayValue` - it actually is not adding too much overhead, only needed in two places (apply/clear)
* Got rid of the `disconnect`, that seems to be causing issues in unit tests.
* No need for eslint disables if we use `myElement.setAttribute('value', someValue);`

```ts
import { Controller } from '@hotwired/stimulus';

/**
 * Adds ability to sync the value or interactions with one input with one
 * or more targeted other inputs.
 *
 * @example
 * <section>
 *   <input type="text" name="title" id="title" />
 *   <input
 *     type="date"
 *     id="event-date"
 *     name="event-date"
 *     value="2025-07-22"
 *     data-controller="w-sync"
 *     data-action="change->w-sync#apply cut->w-sync#clear"
 *     data-w-sync-target-value="#title"
 *   />
 * </section>
 */
export class SyncController extends Controller<HTMLInputElement> {
  static values = {
    delay: { default: 0, type: Number },
    disabled: { default: false, type: Boolean },
    quiet: { default: false, type: Boolean },
    target: String,
  };

  declare delayValue: number;
  declare disabledValue: boolean;
  declare quietValue: boolean;
  declare readonly targetValue: string;

  /**
   * Dispatches an event to all target elements so that they can be notified
   * that a sync has started, allowing them to disable the sync by preventing
   * default.
   */
  connect() {
    this.processTargetElements('start', true);
  }

  /**
   * Allows for targeted elements to determine, via preventing the default event,
   * whether this sync controller should be disabled.
   */
  check() {
    this.processTargetElements('check', true);
  }

  /**
   * Applies a value from the controlled element to the targeted
   * elements.
   */
  apply() {
    this.processTargetElements('apply').forEach((target) => {
      setTimeout(() => {
        target.setAttribute('value', this.element.value);
        if (this.quietValue) return;
        this.dispatch('change', {
          cancelable: false,
          prefix: '',
          target: target as HTMLInputElement,
        });
      }, this.delayValue);
    });
  }

  /**
   * Clears the value of the targeted elements.
   */
  clear() {
    this.processTargetElements('clear').forEach((target) => {
      setTimeout(() => {
        target.setAttribute('value', '');
        if (this.quietValue) return;
        this.dispatch('change', {
          cancelable: false,
          prefix: '',
          target: target as HTMLInputElement,
        });
      }, this.delayValue);
    });
  }

  /**
   * Simple method to dispatch a ping event to the targeted elements.
   */
  ping() {
    this.processTargetElements('ping', false, { bubbles: true });
  }

  /**
   * Returns the non-default prevented elements that are targets of this sync
   * controller.
   */
  processTargetElements(
    eventName: string,
    resetDisabledValue = false,
    options = {},
  ) {
    if (!resetDisabledValue && this.disabledValue) {
      return [];
    }

    const targetElements = [
      ...document.querySelectorAll<HTMLElement>(this.targetValue),
    ];

    const elements = targetElements.filter((target) => {
      const event = this.dispatch(eventName, {
        bubbles: false,
        cancelable: true,
        ...options, // allow overriding some options but not detail & target
        detail: { element: this.element, value: this.element.value },
        target: target as HTMLInputElement, // BUG REF
      });

      return !event.defaultPrevented;
    });

    if (resetDisabledValue) {
      this.disabledValue = targetElements.length > elements.length;
    }

    return elements;
  }
}
```